Griffon11/bjs2 83575

Ana25478 opened the pull request:
Не добавила @Value для месяцев (оставила константу), т.к. не очень поняла, где именно её объявить. Попробую доделать до ретро, но пока в любом случае кидаю ПР

Ana25478 committed:
Init commit with the solution (without an extra complex task)

Ana25478 committed:
Added the 6 months check for repeated recommendations

Ana25478 committed:
Added @Mapping ignore for User field mappings

Ana25478 committed:
Corrected CheckStyle errors

hohlovps95 commented on src/main/java/school/faang/user_service/controller/recommendation/RecommendationController.java:
```diff
@@ -0,0 +1,42 @@
+    private final RecommendationService recommendationService;
+
+    RecommendationDto create(CreateRecommendationDto recommendationDto) {
+        if (recommendationDto.receiverId() != null && !recommendationDto.content().isBlank()) {
```
Мне кажется, что такие проверки загромождают код, проверку на null и isblank лучше вынести в класс DTO  и реализовать через jakarta. Примерно так @NotNull @NotIsBlank content. А в контроллер добавить аннотацию @Validated перед объектом CreateRecommendationDto.

hohlovps95 commented on src/main/java/school/faang/user_service/controller/recommendation/RecommendationController.java:
```diff
@@ -0,0 +1,42 @@
+import java.util.List;
+
+@RequiredArgsConstructor
+@Controller
```
Я думаю, в данном случае, не совсем правильно использовать Controller. Он чаще всего используется с шаблонизаторами типа Thymeleaf. Тут лучше смотрится @RestController. Так как этим функционалом может пользоваться не только Web, но и другие микросервисы.

hohlovps95 commented on src/main/java/school/faang/user_service/controller/recommendation/RecommendationController.java:
```diff
@@ -0,0 +1,42 @@
+@Controller
+public class RecommendationController {
+    private final RecommendationService recommendationService;
+
```
Я бы добавил сюда аннотацию, указывающую на метод Get Post и тд. Например @PostMapping("/recommendation")

hohlovps95 commented on src/main/java/school/faang/user_service/controller/recommendation/RecommendationController.java:
```diff
@@ -0,0 +1,42 @@
+    }
+
+    RecommendationDto update(long recommendationId, UpdateRecommendationDto recommendationDto) {
+        if (!recommendationDto.content().isBlank()) {
```
Про проверки смотри комментарий выше

hohlovps95 commented on src/main/java/school/faang/user_service/service/recommendation/RecommendationServiceImpl.java:
```diff
@@ -0,0 +1,96 @@
+    private final int repeatRecommendationTimeLimit = 6;
+
+
+    @Override
```
К методам, которые вносят в базу какие то изменения лучше добавлять аннотацию @Transactional. Это сделает метод более безопасным. Если в ходе выполнения метода произойдут какие то непредвиденные ситуации, это может сломать базу. Аннотация @Transactional позволяет в случае каких то проблем в методе, откатить изменения в базе, внесенные этим методом.

hohlovps95 commented on src/main/java/school/faang/user_service/service/recommendation/RecommendationServiceImpl.java:
```diff
@@ -0,0 +1,96 @@
+    private final RecommendationRepository recommendationRepository;
+    private final RecommendationMapper recommendationMapper;
+    private final UserContext userContext;
+    private final int repeatRecommendationTimeLimit = 6;
```
Я бы всё таки оставил REPEAT_RECOMMENDATION__TIME_LIMIT, а лучше переместить константу в ямл. И от туда забирать значение.

hohlovps95 reviewed (commented):
Написал свои соображения, посмотри пожалуйста.

Ana25478 committed:
Addressed feedback: added jakarta and additional spring annotations for cleaner code, moved the constant to application.yaml,

zaqqaz226 commented on src/main/java/school/faang/user_service/controller/recommendation/RecommendationController.java:
```diff
@@ -0,0 +1,38 @@
+
+@RequiredArgsConstructor
+@RestController
+public class RecommendationController {
```
Добавь еще аннотацию @RequestMapping("/common-path") . Мое имя общего пути к контроллеру не бери)

Ana25478 replied:
Done (вроде... ещё разбираюсь с аннотациями для сервлетов)

zaqqaz226 commented on src/main/java/school/faang/user_service/controller/recommendation/RecommendationController.java:
```diff
@@ -0,0 +1,38 @@
+    private final RecommendationService recommendationService;
+
+    @PostMapping("/recommendation")
+    RecommendationDto create(@Validated CreateRecommendationDto recommendationDto) {
```
Объяви у всех методов явный модификатор доступа. Так проще читать.

Ana25478 replied:
Готово

zaqqaz226 commented on src/main/java/school/faang/user_service/controller/recommendation/RecommendationController.java:
```diff
@@ -0,0 +1,38 @@
+        return recommendationService.create(recommendationDto);
+    }
+
+    RecommendationDto update(long recommendationId, @Validated UpdateRecommendationDto recommendationDto) {
```
Тут тоже понадобится что-то вроде @PutMapping() или @PatchMapping()

Ana25478 replied:
Добавила

zaqqaz226 commented on src/main/java/school/faang/user_service/controller/recommendation/RecommendationController.java:
```diff
@@ -0,0 +1,38 @@
+        return recommendationService.update(recommendationId, recommendationDto);
+    }
+
+    void delete(long recommendationId) {
```
И тут тоже нужен какой нибудь маппинг, иначе сервлет не знает куда ему идти

Ana25478 replied:
Готово

zaqqaz226 commented on src/main/java/school/faang/user_service/service/recommendation/RecommendationServiceImpl.java:
```diff
@@ -0,0 +1,110 @@
+    @Transactional
+    @Override
+    public RecommendationDto create(CreateRecommendationDto newRecommendationDto) {
+        if (!newRecommendationDto.receiverId().equals(userContext.getUserId())) {
```
Лучше вынести эту проверку в отдельный приватный метод. Для читаемости кода

Ana25478 replied:
Тут немного непонятно. В классе есть 3 проверки (в двух вариациях): отправитель запроса НЕ РАВЕН получателю рекомендации, отправитель запроса РАВЕН автору. Т.е. надо наверно передавать какой-то флаг параметром(нужно равенство или неравенство id для выброса исключения), + в каждой из этих проверок используютмя разные DTO рекорды (т.е. параметром надо будет передавать громоздкое dtoObject.getAuthorId() и т.д.) Т.к. доп метод для проверки предполагает одну строку с if и выброс исключения, выглядит как очень много ненужных движений, а понимаемость кода не факт что улучшится, по крайней мере в том виде, как я это вижу.
Есть ли советы по организыции этого метода? 
П.с. пока оставила как было

zaqqaz226 replied:
А вот как раз у тебя метод create, должен отвечать за создание сущности в БД. А получается, что он еще и проверкой занимается. Если у нас потом проверки разрастутся, то нам все придется вписывать в этот метод (делая его менее читабельным), либо выносить в отдельный метод. Также, если логика данного метода будет расти, то у тебя if и else разъедутся на кучу строк друг от друга, что также снизит читабельность. Плюс, ты правильно говоришь, что все что делает этот Иф, либо проходит на !true, либо исключение и оставляя это в методе, мы просто оставляем лишние строки, которые можно вынести. Проще его вынести в отдельный приватный void метод и все. На вход подай long (newRecommendationDto.receiverId()) и там просто проверь, сходится ли с userContext.getUserId()

zaqqaz226 commented on src/main/java/school/faang/user_service/service/recommendation/RecommendationServiceImpl.java:
```diff
@@ -0,0 +1,110 @@
+    @Override
+    public RecommendationDto create(CreateRecommendationDto newRecommendationDto) {
+        if (!newRecommendationDto.receiverId().equals(userContext.getUserId())) {
+            if (latestRecommendationCheck(newRecommendationDto)) {
```
Выброс исключения, тоже можно перенести в метод latestRecommendationCheck, а тут обойтись без if - else

Ana25478 replied:
Готово

zaqqaz226 commented on src/main/java/school/faang/user_service/service/recommendation/RecommendationServiceImpl.java:
```diff
@@ -0,0 +1,110 @@
+                long newRecommendationId = recommendationRepository.create(authorId,
+                        receiverId, content);
+                return recommendationMapper.toRecommendationDto(recommendationRepository
+                        .findById(newRecommendationId).orElseThrow(EntityNotFoundException::new));
```
Лучше выбрасывать исключение с каким то сообщением, чтобы мы потом могли понять, что пошло не так

Ana25478 replied:
Готово

zaqqaz226 commented on src/main/java/school/faang/user_service/service/recommendation/RecommendationServiceImpl.java:
```diff
@@ -0,0 +1,110 @@
+                return recommendationMapper.toRecommendationDto(recommendationRepository
+                        .findById(newRecommendationId).orElseThrow(EntityNotFoundException::new));
+            } else {
+                throw new ForbiddenException("Latest recommendation limit");
```
Передай в сообщение какие то данные, по которым можно будет понять, что именно пошло не так. По ним будет проще дебажиться

Ana25478 replied:
Готово

zaqqaz226 commented on src/main/java/school/faang/user_service/service/recommendation/RecommendationServiceImpl.java:
```diff
@@ -0,0 +1,110 @@
+    @Override
+    public RecommendationDto update(long recommendationId, UpdateRecommendationDto recommendationDto) {
+        Recommendation recommendationToBeUpdated = recommendationRepository.findById(recommendationId)
+                .orElseThrow(EntityNotFoundException::new);
```
Опять же, лучше если в исключении будет какая нибудь информация

Ana25478 replied:
Готово

Ana25478 committed:
Added exception messages, added Mapping annotations in the Controller class, optimized some check methods

zaqqaz226 commented:
Все отлично, можно делать юнит тесты)

Ana25478 committed:
Added 2 new methods for checking if user matches/ doesn't match operation caller and throw exceptions with custom messaging

zaqqaz226 commented:
Что-то тестов не вижу( Ты точно запушил?

Ana25478 commented:
я думала, в отлельной ветке с тасками для тестов... ещё чуточку позже будет

Ana25478 committed:
Here we go...
Corrected Mapper, Recommendation Filtering method in ServiceImpl, added Filter classes and util filter interface as advised by course materials;
Added test coverage for Service, Controller, new Filter classes, + added FilterInstance classes (manual mocks) as advised by course materials

Ana25478 committed:
Checkstyle correction

Ana25478 committed:
And another one...

Ana25478 committed:
And another one...

Ana25478 committed:
Test logic error correction

Ana25478 committed:
AND ANOTHER CHECKSTYLE

Ana25478 committed:
AND ANOTHER ONE

hohlovps95 commented on src/test/java/school/faang/user_service/filter/RecommendationAuthorFilterTest.java:
```diff
@@ -0,0 +1,96 @@
+    private final RecommendationAuthorFilter recommendationAuthorFilter =
+            new RecommendationAuthorFilter();
+
+    @Test
```
Я бы для дальнейшей читаемости результатов теста добавил бы ещё @DisplayName

Ana25478 replied:
Согласна, добавила

hohlovps95 commented on src/test/java/school/faang/user_service/filter/RecommendationAuthorFilterTest.java:
```diff
@@ -0,0 +1,96 @@
+        long originalAuthorId = 1L;
+        RecommendationFilterDto filterDto =
+                new RecommendationFilterDto(null, originalAuthorId, null);
+        User firstAuthor = User.builder()
```
Билдер, это прям хорошо)

hohlovps95 commented on src/test/java/school/faang/user_service/filter/RecommendationReceiverFilterTest.java:
```diff
@@ -0,0 +1,93 @@
+    }
+
+    @Test
+    public void testApplyNoMatch() {
```
Единственная придирка, я не увидел при негативных тестах, проверки, что у мы выбрасываем действительно правильный эксепшен. То есть не проверяем какой класс эксепшена выстреливает при не валидных данных

Ana25478 replied:
@hohlovps95 Может я не так поняла твой коммент, но по моей логике фильтрации не должно быть исключений. Стрим рекомендаций продолжает фильтроваться с помощью всех имеющихся утильных фильров-классов (сами данные для сравнения берутся из ДТОшки, но они там сами проверяются на валидность). Если никакие записи из базы фильтрам не подойдут - тогда просто вернётся пустой список.
Дай плиз знать, если есть какие идеи/предложения
